### ISSUE TITLE
Provide hints on RunPython migrations to help with multi-database setups

### ISSUE THREAD

**Issue by nkuttler** [wagtail/wagtail#2234]
When using [multiple databases](https://docs.djangoproject.com/en/1.9/topics/db/multi-db/) Wagtail migrations run twice. The second pass fails with

`django.db.utils.IntegrityError: UNIQUE constraint failed: wagtailcore_page.path`

Django docs recommend to check [which database the migrations run against](https://docs.djangoproject.com/en/1.9/howto/writing-migrations/).

I have a trivial fix in the commit below that only prevents this error. It should probably be modified, and documentation needs to be added, but I wanted some feedback first.

**Comment by nkuttler** [comment]
51ef378 demonstrates the problem and leads to all test runs failing.

**Comment by nkuttler** [comment]
I just noticed that Wagtail only supports Django>=1.8 anyway, so the new method of adding a router class would work, and also has the benefit that no migrations at all would be run on the secondary (or any other) databases. I think this would involve modifying all migrations to include something like

    operations = [
        migrations.RunPython(forwards, hints={'target_db': 'default'}),
    ]

where 'default' would be replaced by something like settings.get('WAGTAIL_DB_ALIAS', 'default') or maybe a function call. Alternatively passing the model name might work as well, but the Django documentation is rather terse.

By the way, I'm pretty sure this problem doesn't only happen during unit tests. If you start using Wagtail in an existing project that already has multiple databases, it seems that Wagtail migrations would run on all of those databases. I only added the second database later, after the initial Wagtail migrations were done. Future Wagtail migrations will probably try to run on the secondary database as well, and I have a suspicion that the migrations won't run if they can't find their tables.

**Comment by gasman** [comment]
Hi @nkuttler,
Thanks for investigating this - I was just reading the Django docs for multiple database support, and reaching similar conclusions...

I'd definitely prefer to work with Django's own routing mechanism if possible, rather than creating new Wagtail-specific settings. If I understand correctly, Django is smart enough to automatically select the correct database (`default` in the standard configuration, but may be different if the developer has supplied a custom `DATABASE_ROUTERS` setting) for all standard migration operations (create table, alter table etc) but can't do that for `RunPython` / `RunSQL` operations because it doesn't know that they're working on Wagtail's tables. Based on https://docs.djangoproject.com/en/1.9/topics/db/multi-db/#topics-db-multi-db-hints, it sounds like we should be doing something like:

    operations = [
        # this operation is creating Permission / ContentType objects
        migrations.RunPython(
            create_admin_access_permissions, remove_admin_access_permissions,
            hints={'instance': Permission(name='Some dummy permission')}
        ),
    ]

(The need to pass an instance here seems a bit odd - usually in migrations we'd retrieve the model class using `apps.get_model('auth.Permission')`, but we don't have an `apps` object available here. Maybe I've misunderstood how this is supposed to work - the docs are indeed quite patchy...)

This way, we can avoid having an explicit setting for 'Wagtail's tables belong in database X' - it's determined by the router configuration, as it should be.

**Comment by nkuttler** [comment]
After looking into this further it doesn't look like a Wagtail bug any more. When using multiple databases Django only migrates the default database by default, unless ``--database`` is specified. 

The problem with unit tests is that Django migrates all databases for tests and routed the Wagtail migrations with ``Model.objects.create()`` methods to the default db for each pass, which is correct behavior but lead to the error.

For me a trivial router fixed the problem

    class MyRouter(object):
        def allow_migrate(self, db, app_label, model=None, **hints):
            if db != 'default':
                return False
            return None

In short, people who use multiple databases have to take care of their own routing and (test) migrations, so I'm closing this issue, it was a bug in my own code.

**Comment by gasman** [comment]
Glad you got it sorted! Either way, if I've understood things correctly - it would be a good thing for Wagtail to provide hints on its `RunPython` migrations, because that would allow the default router to work with no extra configuration (for simple setups that put the Wagtail tables in `default`), and give developers extra information to work with if they do need to write more complex routers. So, I'll leave this issue open as a feature request, for anyone interested in taking it on...

**Comment by nkuttler** [comment]
The ``RunPython`` hints are only passed to the ``allow_migrate`` method of the router. The subsequent ``Page.objects.create()`` calls and such in the called function don't inherit them, so the only decision the router makes is if the function should run at all. 

A slightly more complex router example could look like this:

    class MyRouter(object):
        def allow_migrate(self, db, app_label, model=None, **hints):
            if db == 'default' and app_label == 'wagtailimages':
                return False
            return None

This would allow to spread different Wagtail apps across different databases. The ``db`` and ``app_label`` parameters are automatically provided, ``model`` as well but not during ``RunPython`` calls. The model info is probably useful when different models of the same app are stored in different databases, though that use case seems [dubious](https://docs.djangoproject.com/en/1.9/topics/db/multi-db/#no-cross-database-relations), and I guess most Wagtail models have relations between them?

The docs [recommend](https://docs.djangoproject.com/en/1.9/howto/writing-migrations/#data-migrations-and-multiple-databases) to pass a ``model_name`` hint anyway, for example:

    operations = [
        migrations.RunPython(
            initial_data, remove_initial_data,
            hints={'model_name': 'Page'}
        ),
    ]

This would mean to split ``inital_data`` and possibly more migrations so that they only work with one model, e.g.

    operations = [
        migrations.RunPython(
            create_root_page, remove_root_page,
            hints={'model_name': 'Page'}
        ),
        migrations.RunPython(
            create_site, remove_site,
            hints={'model_name': 'Site'}
        ),
    ]

**Comment by gasman** [comment]
Revisiting this in the light of #11184, which has demonstrated that 1) people have been successfully using Wagtail in certain multi-database configurations (until we broke them 🤦), and 2) we need to properly formalise which configurations are and aren't supported (and test the ones that are).

I've now opened #11237 to address this subject. Under that proposal as it currently stands, we would only support configurations where Wagtail's tables exist in the default database and nowhere else, and Wagtail's migrations are suppressed from running in non-default databases. This means that the configuration that triggers this error (in which the Wagtail migrations are allowed to run on multiple databases) would be deemed invalid.

I'm happy to reconsider this if anyone feels that this is excluding valid setups that we ought to be supporting - if so, further feedback would be very welcome at #11237.
